pydantic_ai_slim/pydantic_ai/format_prompt.py

```diff
@@ -79,31 +106,105 @@ def to_xml(self, value: Any, tag: str | None) -> ElementTree.Element:
-            for item in value:  # pyright: ignore[reportUnknownVariableType]
-                item_el = self.to_xml(item, None)
-                element.append(item_el)
+            for n, item in enumerate(value):  # pyright: ignore[reportUnknownVariableType,reportUnknownArgumentType]
+                element.append(self._to_xml(item, None, f'{path}.[{n}]' if path else f'[{n}]'))
         else:
             raise TypeError(f'Unsupported type for XML formatting: {type(value)}')
         return element
 
-    def _mapping_to_xml(self, element: ElementTree.Element, mapping: Mapping[Any, Any]) -> None:
+    def _create_element(self, tag: str, path: str) -> ElementTree.Element:
+        element = ElementTree.Element(tag)
+        if self._fields and path in self._fields:
+            field_repr, field_info = self._fields[path]
+            if self.repeat_field_info or field_repr not in self._parsed_fields:
+                field_attributes = self._extract_attributes(field_info)
+                for k, v in field_attributes.items():
+                    element.set(k, v)
+                self._parsed_fields.add(field_repr)
+        return element
+
+    def _init_fields_info(self):
+        if self.include_field_info and self._fields is None:
+            self._fields = {}
+            self._parse_data_structures(self.data, fields_map=self._fields)
+
+    def _init_element_names(self):
+        if self._element_names is None:
+            self._element_names = {}
+            self._parse_data_structures(self.data, element_names=self._element_names)
+
+    def _mapping_to_xml(
+        self,
+        element: ElementTree.Element,
+        mapping: Mapping[Any, Any],
+        path: str = '',
+    ) -> None:
         for key, value in mapping.items():
             if isinstance(key, int):
                 key = str(key)
             elif not isinstance(key, str):
                 raise TypeError(f'Unsupported key type for XML formatting: {type(key)}, only str and int are allowed')
-            element.append(self.to_xml(value, key))
+            element.append(self._to_xml(value, key, f'{path}.{key}' if path else key))
+
+    @classmethod
+    def _parse_data_structures(
+        cls,
+        value: Any,
+        element_names: dict[str, str] | None = None,
+        fields_map: dict[str, tuple[str, FieldInfo | ComputedFieldInfo]] | None = None,
+        path: str = '',
+    ):
+        """Parse data structures as dataclasses or Pydantic models to extract element names and attributes."""
+        if value is None or isinstance(value, (str, int, float, date, bytearray, bytes, bool)):
+            return
+        elif isinstance(value, Mapping):
+            for k, v in value.items():  # pyright: ignore[reportUnknownVariableType]
+                cls._parse_data_structures(v, element_names, fields_map, f'{path}.{k}' if path else f'{k}')
+        elif is_dataclass(value) and not isinstance(value, type):
+            if element_names is not None:
```

Comment:
Could we give `self._element_names` a default value of `{}` and always wriet directly into that instead of checking for `None` and passing `element_names` around as an arg?

Same for `fields_map`